Implemented hibp. Solves #2476

appieschot opened the pull request:
Implemented hibp. Solves #2476

appieschot commented:
Hmm, confusing.. Tests are working on WIndows locally but seem to fail here 😵

appieschot commented:
Okey @waldekmastykarz added a single log line to see what happens and (while not sure why it doesn't do work) I am one step further:
- The line `console.log(loggerLogSpy.getCall(0).args); // eslint-disable-line no-console` returns `TypeError: Cannot read property 'args' of null` (if I do not add a null check) 
- So it looks like the loggerspy is never populated with results ... I cannot explain why that would be the case on windows though .. Must be me screwing up somewhere..

Any other idea's :D?

waldekmastykarz commented:
Not off the top of my head. Let me have a look at it and get back to you.

waldekmastykarz commented:
It's odd: locally, all tests are passing for me 🤔

waldekmastykarz commented:
I wonder if it could be related to the fact that we're trying to assert stub call with an array with an object that's different from the actual object sent to the log (the difference is in the fact that we're creating two separate instances of what's inside the array so the reference isn't the same). We could test if this is the case, by changing the assert to compare JSON.stringify values of args of the last call to logger and what we assume it should be, like:

```
assert.strictEqual(loggerLogSpy.lastCall.args[0], JSON.stringify([{ "Name": "Adobe" }]));
```

appieschot commented:
> I wonder if it could be related to the fact that we're trying to assert stub call with an array with an object that's different from the actual object sent to the log (the difference is in the fact that we're creating two separate instances of what's inside the array so the reference isn't the same). We could test if this is the case, by changing the assert to compare JSON.stringify values of args of the last call to logger and what we assume it should be, like:
> 
> ```
> assert.strictEqual(JSON.stringify(loggerLogSpy.lastCall.args[0]), JSON.stringify([{ "Name": "Adobe" }]));
> ```

Weird thing is that if I run that test I get an error that the loggerLogSpy.lastCall === null (and thus has no args) 😵

waldekmastykarz commented:
Perhaps it would help to check if the command returns an error, by adding an extra assert: `assert.strictEqual(err, undefined)`

appieschot commented:
One million commits later 🥳, turns out my bad; implemented the `AnonymouysCommand` from the wrong place, that in itself did not throw an error, but does explain why it wasn't working. After fixing that I got a bunch of linting errors (forgot a `,` and added a `space` on a place that was not allowed). After fixing those it turns out I missed a test case for logging an error and additional validation ... What a mess I have made 😂 All works good, pay attention to the fact that it now has 8 commits  #sorry

waldekmastykarz commented on src/m365/aad/commands/user/user-hibp.ts:
```diff
@@ -0,0 +1,84 @@
+interface Options extends GlobalOptions {
+  userName: string;
+  apiKey: string;
+  domain?: string
```
Since `domain` is optional, we should track if it's used in the telemetry

waldekmastykarz reviewed (approved):
Awesome! Merged with just one small fix that I've done when merging the PR 👏

waldekmastykarz commented:
Merged manually. Thank you! 👏
